### PR TITLE
[Bug Fix] Populating SharedWorkSpace Container for Initial File Load

### PR DESCRIPTION
**Pull request by ritikramuka**
This pull request primarily focuses on improving the `loadContainer` function in the `webworker.js` file and refining the handling of entity information in the `extension.ts` file. The changes in `webworker.js` include fixing a typo in a variable name, adding a new function, and adjusting the logic to handle initial state. In `extension.ts`, the changes involve better management of active text editor changes, workspace state changes, and the creation of a web worker instance.

Changes to `webworker.js`:

* [`src/web/client/common/worker/webworker.js`](diffhunk://#diff-8e0adceeebffaef7b18326a30ca124d6e2d8448e6732876fdf21ee42b14897c1L81-R81): Corrected a typo in a variable name from `intital` to `initial` and changed its initial value to `true`.
* [`src/web/client/common/worker/webworker.js`](diffhunk://#diff-8e0adceeebffaef7b18326a30ca124d6e2d8448e6732876fdf21ee42b14897c1R96-R144): Added a new function `getUserIdByConnectionId` inside `loadContainer` to get user details based on connection id. Adjusted the logic to handle initial state and post messages accordingly. [[1]](diffhunk://#diff-8e0adceeebffaef7b18326a30ca124d6e2d8448e6732876fdf21ee42b14897c1R96-R144) [[2]](diffhunk://#diff-8e0adceeebffaef7b18326a30ca124d6e2d8448e6732876fdf21ee42b14897c1L126-R169) [[3]](diffhunk://#diff-8e0adceeebffaef7b18326a30ca124d6e2d8448e6732876fdf21ee42b14897c1R210-R212)

Changes to `extension.ts`:

* [`src/web/client/extension.ts`](diffhunk://#diff-1429bea43439e067ae9056ff3f9f71203533f7d3ca168f0d5469aa1b9a6e0360R147): Added a function to handle active text editor changes and send messages to a web worker for co-presence. [[1]](diffhunk://#diff-1429bea43439e067ae9056ff3f9f71203533f7d3ca168f0d5469aa1b9a6e0360R147) [[2]](diffhunk://#diff-1429bea43439e067ae9056ff3f9f71203533f7d3ca168f0d5469aa1b9a6e0360L198-R204) [[3]](diffhunk://#diff-1429bea43439e067ae9056ff3f9f71203533f7d3ca168f0d5469aa1b9a6e0360L273-R316) [[4]](diffhunk://#diff-1429bea43439e067ae9056ff3f9f71203533f7d3ca168f0d5469aa1b9a6e0360L300-L314) [[5]](diffhunk://#diff-1429bea43439e067ae9056ff3f9f71203533f7d3ca168f0d5469aa1b9a6e0360L333-R354) [[6]](diffhunk://#diff-1429bea43439e067ae9056ff3f9f71203533f7d3ca168f0d5469aa1b9a6e0360R425-R432)
* [`src/web/client/extension.ts`](diffhunk://#diff-1429bea43439e067ae9056ff3f9f71203533f7d3ca168f0d5469aa1b9a6e0360L333-R354): Corrected a typo in a function name from `processWillStartCollaboartion` to `processWillStartCollaboration`.
* [`src/web/client/extension.ts`](diffhunk://#diff-1429bea43439e067ae9056ff3f9f71203533f7d3ca168f0d5469aa1b9a6e0360R29): Added the `getFileRootWebPageId` function to get the root web page id for a file. This function was used in various parts of the code to improve the handling of entity information. [[1]](diffhunk://#diff-1429bea43439e067ae9056ff3f9f71203533f7d3ca168f0d5469aa1b9a6e0360R29) [[2]](diffhunk://#diff-1429bea43439e067ae9056ff3f9f71203533f7d3ca168f0d5469aa1b9a6e0360L273-R316) [[3]](diffhunk://#diff-1429bea43439e067ae9056ff3f9f71203533f7d3ca168f0d5469aa1b9a6e0360R425-R432) [[4]](diffhunk://#diff-02b9a80bec87b6e3c39911c83e2d679402dbe0c4e87ceddf6bf13bd5f3b6846cR32-R36) [[5]](diffhunk://#diff-8c87e5f3eec6c61061b920113621fd5edb0ad12e511628c798dc63f5b611673dL10-R10) [[6]](diffhunk://#diff-8c87e5f3eec6c61061b920113621fd5edb0ad12e511628c798dc63f5b611673dL27-R31)

These changes aim to improve the efficiency of the code and make it easier to manage and understand.

### CODE REVIEW

**Review by BidishaMS — APPROVED** [review]

File: src/web/client/common/worker/webworker.js [hunk 1]
```diff
@@ -163,20 +138,79 @@ async function loadContainer(config, swpId, entityInfo) {
...
+                    // aadObjectId is the unique identifier for a user
+                    await self.postMessage({
+                        type: "client-data",
+                        userId: user.aadObjectId,
+                        userName: user.userName,
+                        containerId: swpId,
+                        entityId: userEntityIdArray,
+                    });
+
+                    await self.postMessage({
```
Comment (tyaginidhi) [inline src/web/client/common/worker/webworker.js:197]: We do not need to await telemetry

File: src/web/client/common/worker/webworker.js [hunk 2]
```diff
@@ -163,20 +138,79 @@ async function loadContainer(config, swpId, entityInfo) {
...
+                        .get();
+
+                    userConnections.forEach((connection) => {
+                        userEntityIdArray.push(
+                            connectionIdInContainer.get(connection.id)
+                        );
+                    });
+
+                    // aadObjectId is the unique identifier for a user
+                    await self.postMessage({
```
Comment (tyaginidhi) [inline src/web/client/common/worker/webworker.js:189]: Do we really need to await this one?

File: src/web/client/common/worker/webworker.js [hunk 3]
```diff
@@ -163,20 +138,79 @@ async function loadContainer(config, swpId, entityInfo) {
...
+                        entityId: userEntityIdArray,
+                    });
+
+                    await self.postMessage({
+                        type: "telemetry-info",
+                        eventName: "webExtensionWebWorkerGetUserIdByConnectionIdSuccess",
+                        userId: user.aadObjectId,
+                    });
+                } catch (error) {
+                    await self.postMessage({
```
Comment (tyaginidhi) [inline src/web/client/common/worker/webworker.js:203]: No need to await telemetry

File: src/web/client/common/worker/webworker.js [hunk 4]
```diff
@@ -163,20 +138,79 @@ async function loadContainer(config, swpId, entityInfo) {
...
+                        type: "telemetry-error",
+                        methodName: "webWorker valueChanged",
+                        errorMessage: error?.message,
+                        error: error,
+                    });
+                }
+            } else {
+                initial = false;
```
Comment (tyaginidhi) [inline src/web/client/common/worker/webworker.js:211]: What are we trying to do here with initial? Doesn't look optimal - can you discuss this approach once with someone in team?

File: src/web/client/extension.ts [hunk 5]
```diff
@@ -270,35 +272,11 @@ export function processWorkspaceStateChanges(context: vscode.ExtensionContext) {
                     const entityInfo: IEntityInfo = {
                         entityId: getFileEntityId(document.uri.fsPath),
                         entityName: getFileEntityName(document.uri.fsPath),
-                        rootWebPageId: WebExtensionContext.entityDataMap.getEntityMap.get(getFileEntityId(document.uri.fsPath))?.rootWebPageId as string
+                        rootWebPageId: getFileRootWebPageId(document.uri.fsPath),
```
Comment (tyaginidhi) [inline src/web/client/extension.ts:275]: This can be null value now - are we handling that scenario

File: src/web/client/extension.ts [hunk 6]
```diff
@@ -314,6 +292,49 @@ export function processWorkspaceStateChanges(context: vscode.ExtensionContext) {
...
+                    sendingMessageToWebWorkerForCoPresence(entityInfo);
+                }
+
+                WebExtensionContext.quickPickProvider.refresh();
+            }
+        })
+    );
+}
+
+function sendingMessageToWebWorkerForCoPresence(entityInfo: IEntityInfo) {
```
Comment (tyaginidhi) [inline src/web/client/extension.ts:316]: Can we move this function to copresence utility file - this does not belong in entry file for extension

File: src/web/client/utilities/fileAndEntityUtil.ts [hunk 7]
```diff
@@ -29,6 +29,11 @@ export function getFileEntityName(fileFsPath: string) {
         ?.entityName as string ?? WebExtensionContext.getVscodeWorkspaceState(fileFsPath)?.entityName as string;
 }
 
+export function getFileRootWebPageId(fileFsPath: string) {
+    const entityId = getFileEntityId(fileFsPath);
+    return WebExtensionContext.entityDataMap.getEntityMap.get(entityId)?.rootWebPageId as string ?? WebExtensionContext.getVscodeWorkspaceState(fileFsPath)?.rootWebPageId as string;
```
Comment (tyaginidhi) [inline src/web/client/utilities/fileAndEntityUtil.ts:34]: Can we simplify this code for readability - and easy debugging- way too nested code

Comment (ritikramuka) [reply src/web/client/utilities/fileAndEntityUtil.ts:34]: I'm following the same convention followed in this file

File: src/web/client/extension.ts (hunk 6 above)
Comment (ritikramuka) [reply src/web/client/extension.ts:316]: done!

File: src/web/client/extension.ts (hunk 5 above)
Comment (ritikramuka) [reply src/web/client/extension.ts:275]: yeah, we should handle that.

File: src/web/client/common/worker/webworker.js (hunk 4 above)
Comment (ritikramuka) [reply src/web/client/common/worker/webworker.js:211]: @BidishaMS, lets discuss this

Comment (BidishaMS) [reply src/web/client/common/worker/webworker.js]: Reviewed and discussed different approach with @Ritik. He has implemented the same

File: src/web/client/utilities/fileAndEntityUtil.ts [hunk 8]
```diff
@@ -29,6 +29,17 @@ export function getFileEntityName(fileFsPath: string) {
         ?.entityName as string ?? WebExtensionContext.getVscodeWorkspaceState(fileFsPath)?.entityName as string;
 }
 
+export function getFileRootWebPageId(fileFsPath: string) {
+    const entityId = getFileEntityId(fileFsPath);
+    return (
+        (WebExtensionContext.entityDataMap.getEntityMap.get(entityId)
```
Comment (tyaginidhi) [inline src/web/client/utilities/fileAndEntityUtil.ts:35]: Do add unit tests to make sure this logic is robust

Comment (ritikramuka) [reply src/web/client/utilities/fileAndEntityUtil.ts:35]: sure! will add in a new PR

**Pull request merged**